File: python/2015/day05a/day05a.py
#!/usr/bin/env python3

import re
import sys

def usage():
    progname = sys.argv[0]
    print(f'usage: {progname} <input file>')
    sys.exit(1)

def prop1(line):
    return len(re.findall(r'[aeiou]', line)) >= 3

def prop2(line):
    return bool(re.search(r'(.)\1', line))

def prop3(line):
    return not any(map(lambda x: x in line, ['ab', 'cd', 'pq', 'xy']))

def process(filename):
    counter = 0
    with open(filename, 'r') as infile:
        for line in infile:
            if prop1(line) and prop2(line) and prop3(line):
                counter += 1
    return counter

def main():
    if len(sys.argv) < 2:
        usage()
    filename = sys.argv[1]
    result = process(filename)
    print(f'result = {result}')

if __name__ == '__main__':
    main()

File: ruby/2015/day05a/day05a.rb
#!/usr/bin/env ruby

require 'set'

def usage
    progname = File.basename($0)
    puts "usage: #{progname} <input file>"
    exit 1
end

def prop1?(str)
    vowels = str.scan(/[aeiou]/)
    vowels.length >= 3
end

def prop2?(str)
    !!(str =~ /(.)\1/)
end

def prop3?(str)
    !(str =~ /(ab|cd|pq|xy)/)
end

def process(filename)
    counter = 0
    File.foreach(filename) do |line|
        counter += 1 if prop1?(line) && prop2?(line) && prop3?(line)
    end
    counter
end

def main
    usage if ARGV.length < 1
    filename = ARGV[0]
    result = process(filename)
    puts "result = #{result}"
end

main if __FILE__ == $0
